### PR TITLE
Enhancements and fixes

### LINKED ISSUE #37: Reduce trigonometric calculations in angular_hit.
During our presentation, Matt was asking about calls to expensive functions, and other optimizations we can conduct. Below I present an optimization in the ```angular_hit``` function. I confess I am hesitant to implement any optimizations yet, at least not until we have a working 3-dimensional implementation.

Right now in ```angular_hit```, we have the following:
```
delta_theta = 2 * pi / num_angular_sections;
interval_theta = [current_voxel_ID_theta * delta_theta, (current_voxel_ID_theta + 1) * delta_theta];

xmin = cos(min(interval_theta));
xmax = cos(max(interval_theta));
ymin = sin(min(interval_theta));
ymax = sin(max(interval_theta));

if (single(tan(min(interval_theta))) == ray_direction(2)/ray_direction(1) ...
        && single(tan(max(interval_theta))) == ray_direction(2)/ray_direction(1))
        fprintf("parallel");
        ...
end
```

In a single call to ```radial_hit```, this ends up being the following :
**function | #**
```cos``` | 2 calls
```sin``` | 2 calls
``` tan``` | 2 calls
We also calculate ```delta_theta``` (float multiplication, float division) and ```interval_theta``` (2 float multiplications).

Multiply this by N, where N is the number of angular section traversals.

We can reduce this to 1 by adding the following calculations to the initialization phase in ```polarCoordinateTraversal.m```:
```
delta_theta = 2 * pi / num_angular_sections;
interval_thetas = zeros(num_angular_sections);
interval_theta_cos = zeros(num_angular_sections);
interval_theta_sin = zeros(num_angular_sections);
interval_theta_tan = zeros(num_angular_sections);

% Calculate cosine min/max, sine min/max for angular sections.
for k = 1 : num_angular_sections
    i = (k - 1) * delta_theta;  % Recall: theta goes from 0 to N - 1, but MATLAB begins at 1.
    interval_thetas(k) = i;
    interval_theta_cos(k) = cos(i);
    interval_theta_sin(k) = sin(i);
    interval_theta_tan(k) = sin(i);
end
```

And then, in ```angular_hit```, we need only to reference these pre-calculated numbers:
```
% Note: must be careful about off-by-one error.
voxel_one = current_voxel_ID_theta + 1;
voxel_two = current_voxel_ID_theta + 2;
interval_theta = [interval_thetas(voxel_one), interval_thetas(voxel_two)];

xmin = min(interval_theta_cos(voxel_one),  interval_theta_cos(voxel_two));
xmax = max(interval_theta_cos(voxel_one),  interval_theta_cos(voxel_two));
ymin = min(interval_theta_sin(voxel_one),  interval_theta_sin(voxel_two));
ymax = max(interval_theta_sin(voxel_one),  interval_theta_sin(voxel_two));

y_over_x = ray_direction(2) / ray_direction(1);
is_parallel_min = single(min(interval_theta_tan(voxel_one), interval_theta_tan(voxel_two) )) == y_over_x;
is_parallel_max = single(max(interval_theta_tan[voxel_one], interval_theta_tan(voxel_two) )) == y_over_x;
if (is_parallel_min && is_parallel_max)
   fprintf("parallel\n");
   ...
end
```

This cuts out a lot of the meat in our hot loop (see exact numbers above), at the cost of a computationally more expensive start-up time. Given many rays, this will be worth it though.

With regards to ```atan2``` (asked about in the presentation), we call this once per ray in the initialization phase as well.

### LINKED ISSUE #39: Edge case with ray beginning at a theta boundary test.
```
% Ray begins at a theta boundary (and goes in orthogonal directions).
function testBeginAtThetaBoundaryInOrthogonal(testCase)
    fprintf("Ray begins at a theta boundary (and goes in orthogonal directions\n")
    min_bound = [0.0, 0.0];
    max_bound = [30.0, 30.0];
    ray_origin = [10.0, 15.0];
    ray_direction = [0.0, 7.0];
    circle_center = [15.0, 15.0];
    circle_max_radius = 10.0;
    num_radial_sections = 3;
    num_angular_sections = 4;
    t_begin = 0.0;
    t_end = 20.0;
    verbose = false;
    
    [rVoxels, thetaVoxels] = polarCoordinateTraversal(min_bound, max_bound, ray_origin, ray_direction, ...
        circle_center, circle_max_radius, num_radial_sections, num_angular_sections, t_begin, t_end, verbose);
        expected_rVoxels     = [2,1];
        expected_thetaVoxels = [1,1];
        
        verifyEqual(testCase, rVoxels, expected_rVoxels);
        verifyEqual(testCase, thetaVoxels, expected_thetaVoxels);
end
```

Here, we get radial_voxels: ```[2, 2, 1]```
when we should be expecting 
radial_voxels: ```[2, 1]```

This means the theta voxel begins in  theta voxel 2, and traverses into 1, which causes an extra hit. 

We seem to agree that as long as we aren't intersecting with objects that aren't in the ray's path, and we aren't causing any unnecessary extension (or deletion) in the chord line, we should be ok.

### LINKED ISSUE #40: Edge case with along theta boundary test
```
% Ray goes along a theta boundary.
function testRayGoesAlongThetaBoundary(testCase)
    fprintf("Ray goes along a theta boundary\n")
    min_bound = [0.0, 0.0];
    max_bound = [30.0, 30.0];
    ray_origin = [25.0, 15.0];
    ray_direction = [-1.0, 0.0];
    circle_center = [15.0, 15.0];
    circle_max_radius = 10.0;
    num_radial_sections = 3;
    num_angular_sections = 4;
    t_begin = 0.0;
    t_end = 20.0;
    verbose = false;
    
    [rVoxels, thetaVoxels] = polarCoordinateTraversal(min_bound, max_bound, ray_origin, ray_direction, ...
        circle_center, circle_max_radius, num_radial_sections, num_angular_sections, t_begin, t_end, verbose);
        expected_rVoxels     = [1,2,3,3,2,1];
        expected_thetaVoxels = [0,0,0,1,1,1];
        
        verifyEqual(testCase, rVoxels, expected_rVoxels);
        verifyEqual(testCase, thetaVoxels, expected_thetaVoxels);
end
```

The actual_thetaVoxels is instead ```[0,0,0,2,2,2]```. This is making a jump to 2 instead of 1 since its along the theta boundary. This issue is just to ensure we are aware in case it comes up again in the future.

### LINKED ISSUE #73: Fix theta voxel initialization phase.
## Issue found with small voxel count:
78264483-38506080-74d1-11ea-93e9-29a639f36a89

```
thetaVoxels =  1     2     2     2     2     2     2     2     2     2     3     
               3     3     3     3     4     4     5     6     7     7     8     
               8     8     9     9     9     9     9     9     9     9     10    
               10    10    10
```
The algorithm produces thetaVoxel as being 1 during initialization phase (using both old version with atan2 and new version with no trigonometric function), yet this image clearly depicts it being 2.
If I double the num_angular_sections from 20 to 40, it is off by one again.
My guess is that our current calculations are not taking into account the position along the ray at first entry. So, we need:

```
if abs(ray_origin - circle_center) < tol 
    % If the ray starts at the origin, we need to perturb slightly along its 
    % path to find the correct angular voxel.
    pert_t = 0.1;
    pert_x = ray_origin(1) + ray_direction_x * pert_t;
    pert_y = ray_origin(2) + ray_direction_y * pert_t;
    a = circle_center(1) - pert_x;
    b = circle_center(2) - pert_y;
else
    % Otherwise, we want the theta voxel where the ray intersects with the first radial boundary. 
    ray_circle_intersection_x = ray_origin(1) + ray_direction_x * t1; % t1 calculated in initialization phase
    ray_circle_intersection_y = ray_origin(2) + ray_direction_y * t1;
    a = circle_center(1) - ray_circle_intersection_x;
    b = circle_center(2) - ray_circle_intersection_y;
end
```
We then need to update t to be ```t = max(t_begin, t1);```, so the first angular_hit does not go back.
And it works! At least for this case. We're assuming here though that it'll hit a radial boundary before it traverses a theta boundary. So what we may want to do is have the following:

```
if abs(ray_origin - circle_center) < tol 
    % If the ray starts at the origin, we need to perturb slightly along its 
    % path to find the correct angular voxel.
    pert_t = 0.1;
    pert_x = ray_origin(1) + ray_direction_x * pert_t;
    pert_y = ray_origin(2) + ray_direction_y * pert_t;
    a = circle_center(1) - pert_x;
    b = circle_center(2) - pert_y;
elseif the ray begins inside the circle, do what we were originally doing.
else
    % Otherwise, we want the theta voxel where the ray intersects with the first radial boundary. 
    ray_circle_intersection_x = ray_origin(1) + ray_direction_x * t1; % t1 calculated in initialization phase
    ray_circle_intersection_y = ray_origin(2) + ray_direction_y * t1;
    a = circle_center(1) - ray_circle_intersection_x;
    b = circle_center(2) - ray_circle_intersection_y;
end
```
Its important to note that all of our tests pass with the first code segment, which means we need to add a test that hits this case.

New PR to follow soon.

_Originally posted by @cgyurgyik in https://github.com/spherical-volume-rendering/algorithm-team-collaboration/pull/58#issuecomment-607901974_

### PR DESCRIPTION
**Pull request by ak-2485**
1. Remove atan2 from angular init (Fixes #37, Fixes #73 )
2. Remove linear algebra from angular hit
3. Remove sine/cosine from angular hit (Fixes #37 )
4. Add function for testing equivalence of floats
5. Update tests (Fixes #40, Fixes #39 )

### CODE REVIEW

File: matlab/polar-traversal/angular_hit.m [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
-    num_angular_sections, circle_center, t, verbose)
+    num_angular_sections, circle_center, pointArray, t, t_end, verbose)
 % Determines whether an angular hit occurs for the given ray.
 % Input:
 %    ray_origin: vector of the origin of the ray in cartesian coordinate.
 %    ray_direction: vector of the direction of the ray in cartesian coordinate.
 %    current_voxel_ID_theta: the (angular) ID of current voxel.
 %    num_angular_sections: number of total angular sections on the grid.
 %    circle_center: The center of the circle.
+%    pointArray:
```
Comment (cgyurgyik) [inline matlab/polar-traversal/angular_hit.m:10]: Document please.

File: matlab/polar-traversal/approximatelyEqual.m [hunk 2]
```diff
@@ -0,0 +1,8 @@
+function [check] = approximatelyEqual(a, b, absEpsilon, relEpsilon)
+% https://www.learncpp.com/cpp-tutorial/relational-operators-and-floating-point-comparisons/
+    diff = abs(a-b);
+    if diff <= absEpsilon
+        check = true;
```
Comment (cgyurgyik) [inline matlab/polar-traversal/approximatelyEqual.m:5]: Did you mean to return after this?

**Comment by cgyurgyik** [comment]
General q: Why do you use 0,1s instead of true, false? 
i.e., intersect_max, intersect_min, change_r, ...

File: matlab/polar-traversal/angular_hit.m [hunk 3]
```diff
@@ -17,89 +19,145 @@
...
+    if  approximatelyEqual(t_min,t_max,1e-12,1e-8) && ...
+            t < t_min && t_min < t_end && ...
+            ~approximatelyEqual(t,t_min,1e-12,1e-8)
+        tMaxTheta = t_max;
+         if verbose
+            fprintf("hit origin t\n")
+        end
+        if ~approximatelyEqual(ray_direction(2),0.0,1e-12,1e-8)
+            % change in theta is dependent on the slope of the line
+            tStepTheta = - num_angular_sections/2;
```
Comment (cgyurgyik) [inline matlab/polar-traversal/angular_hit.m:125]: Does this also work for odd number of angular sections?

File: matlab/polar-traversal/approximatelyEqual.m (hunk 2 above)
Comment (ak-2485) [reply matlab/polar-traversal/approximatelyEqual.m:5]: Probably not; need conditional for odd case.

File: matlab/polar-traversal/angular_hit.m (hunk 3 above)
Comment (ak-2485) [reply matlab/polar-traversal/angular_hit.m:125]: Probably not; need conditional for odd case.

File: matlab/polar-traversal/approximatelyEqual.m (hunk 2 above)
Comment (ak-2485) [reply matlab/polar-traversal/approximatelyEqual.m:5]: yes!
